### DISCUSSION TITLE
Feature request: JSON extract array

### DISCUSSION THREAD

**Discussion by dmpetrov** [duckdb/duckdb#4599] [General]
I was unable to find a way of working with JSON arrays of unknown size.

An example: I'm looking for a solution of extracting `LabelName` from an arbitrary size list `detections`.

```json
"detections": [
    {
        "Confidence": 1,
        "LabelName": "/m/0b_rs",
        "Source": "xclick"
    },
    {
        "Confidence": 1,
        "LabelName": "/m/0bt9lr",
        "Source": "xclick"
    }
]
```

In BigQuery it will look like (see `JSON_QUERY_ARRAY`):
```sql
CREATE TABLE meta (
    name    STRING,
    data    JSON
)
....

SELECT name, data, '/m/0bt9lr' IN UNNEST(labels) AS is_class_presented
FROM (
    SELECT
        ROW_NUMBER() OVER() AS num,
        ANY_VALUE(name) AS name,
        ANY_VALUE(data) AS data,
        ARRAY_AGG(JSON_VALUE(detection.LabelName)) AS labels
    FROM myds.meta, UNNEST(JSON_QUERY_ARRAY(data.detections)) AS detection
    ORDER BY num
)
```

To unnest the detections, something like JSON_QUERY_ARRAY is needed.

Given DuckDB's naming, I'd propose `json_extract_array()` as the name for this feature.

References:
- `JSON_QUERY_ARRAY()` in BigQueries.
- `json_array_elements` in PostgeSQL.
- Snowflake can UNNEST/FLATTEN json array right from JSON field which looks very nice.

**Comment by koenvo** [comment]
I believe this can be solved using `unnest(from_json(data.detections, '["json"]'))`. See also https://discord.com/channels/909674491309850675/921073327009853451/998937990913077288

**Reply by dmpetrov** [reply]
Thank you! `from_json()` did help.

It might worth to add a bit more information about this to docs. Examples - even better.

**Reply by venkyvb** [reply]
I am curious about the above solution. Don't seem to have access to the discord channel, would be great if you could provide an example in this thread.

**Reply by Mause** [reply]
> I am curious about the above solution. Don't seem to have access to the discord channel, would be great if you could provide an example in this thread.

The example from discord is:

```sql
select unnest(from_json(column_with_json_array, '["json"]')) from my_table
```

**Reply by dmpetrov** [reply]
> The example from discord is:

Yes. For the initial example the solution might look like:

```sql
SELECT
    FIRST(name) AS name,
    FIRST(data) AS data,
    list_contains(LIST(v1), '/m/0bt9lr') AS is_class_presented
FROM (
    SELECT ROW_NUMBER() OVER() AS _num,
        name, data,
        UNNEST(from_json(data -> 'detections', '["json"]')) ->> 'LabelName' AS v1
    FROM meta
)
GROUP BY _num
```

**Reply by dmpetrov** [reply]
> I believe this can be solved using `unnest(from_json(data.detections, '["json"]'))`

It stop working after update to 0.5.0 😬 
I'd appreciate it if you can share any ideas on how it can be fixed.

**Reply by Alex-Monahan** [reply]
What error do you see now?

**Reply by dmpetrov** [reply]
```
duckdb.BinderException: Binder Error:  Invalid LIST argument to from_json!
```

**Reply by dmpetrov** [reply]
Ok. I was able to fix this but still don't fully understand the reason.

```
$ SELECT from_json(data -> 'detections', '["json"]') FROM meta
duckdb.BinderException: Binder Error:  Invalid LIST argument to from_json!

$ SELECT from_json(data ->> 'detections', '["json"]') FROM meta
<<works!>>
```

In 0.4.0 the first query worked just fine.

**Comment by dmpetrov** [comment]
I just realized that `json_contains()` might be a better solution for the problem that I have. Like:
```sql
SELECT * FROM meta
WHERE json_contains(json_extract(j, '$. detections[*]. LabelName'),json_array('/m/0bt9lr'));
```

I started a discussion here: https://discord.com/channels/909674491309850675/921073327009853451/1018565171985268766

**Reply by Alex-Monahan** [reply]
Maybe a combination of EXISTS and a list Lambda function would work?
https://duckdb.org/docs/sql/functions/nested#lambda-functions

**Reply by dmpetrov** [reply]
Thank you, it worked out!

```sql
SELECT
   list_contains(list_transform(from_json(data ->> 'detections', '["json"]'),
                                x -> json_extract_string(json(x), 'LabelName') ),
                 '/m/0bt9lr')
FROM meta
```
